ircd/s_serv.c

```diff
@@ -510,26 +512,37 @@ burst_modes_TS6(struct Client *client_p, struct Channel *chptr,
 {
 	rb_dlink_node *ptr;
 	struct Ban *banptr;
+	char bmask[BANLEN];
 
-	send_multiline_init(client_p, " ", ":%s BMASK %ld %s %c :",
-			me.id,
-			(long)chptr->channelts,
-			chptr->chname,
-			flag);
+	sprintf(buf, ":%s BMASK %ld %s %c",
+		me.id,
+		(long)chptr->channelts,
+		chptr->chname,
+		flag);
+
+	send_multiline_init(client_p, " ", "%s :", buf);
 
 	RB_DLINK_FOREACH_PREV(ptr, list->tail)
 	{
 		banptr = ptr->data;
 
 		if (banptr->forward)
-			send_multiline_item(client_p, "%s$%s",
-					banptr->banstr,
-					banptr->forward);
+			sprintf(bmask, "%s$%s", banptr->banstr, banptr->forward);
+		else
+			strcpy(bmask, banptr->banstr);
+
+		if (IsCapable(client_p, CAP_EBMASK))
+			sendto_one(client_p, "%s %s %ld %s",
+				buf,
+				bmask,
+				(long)banptr->when,
+				banptr->who);
 		else
-			send_multiline_item(client_p, "%s", banptr->banstr);
+			send_multiline_item(client_p, "%s", bmask);
 	}
 
-	send_multiline_fini(client_p, NULL);
+	if (!IsCapable(client_p, CAP_EBMASK))
+		send_multiline_fini(client_p, NULL);
```

Comment:
Code-wise, it might be better to have a separate `burst_modes_EBMASK` and leave `burst_modes_TS6` as is, to avoid the complicated logic with multiple `CAP_EBMASK` checks. I think it is acceptable to iterate over the ban list twice; the optional `%s$%s` construction might be extracted to a function.

This applies particularly if it is planned to deprecate non-EBMASK entirely eventually since `burst_modes_TS6` can be removed at that point without untangling any logic.